compiler/cmdlinehelper.nim

```diff
@@ -0,0 +1,83 @@
+## Helpers for binaries that use compiler passes, eg: nim, nimsuggest, nimfix
+
+# TODO: nimfix should use this; currently out of sync
+
+import
+  compiler/[options, idents, nimconf, scriptconfig, extccomp, commands, msgs, lineinfos, modulegraphs, condsyms],
+  std/os
+
+type ProgBase* = ref object of RootObj
+  name*: string
+  suggestMode*: bool
+
+method processCmdLine(self: ProgBase, pass: TCmdLinePass, cmd: string; config: ConfigRef) {.base.} =
+  doAssert false
+
+method mainCommand(self: ProgBase, graph: ModuleGraph) {.base.} =
+  doAssert false
+
+proc initDefinesProg*(self: ProgBase, conf: ConfigRef) =
+  condsyms.initDefines(conf.symbols)
+  defineSymbol conf.symbols, self.name
+
+proc processCmdLineAndProjectPath*(self: ProgBase, conf: ConfigRef) =
+  self.processCmdLine(passCmd1, "", conf)
+  if conf.projectName == "-":
+    conf.projectName = "stdinfile"
+    conf.projectFull = "stdinfile"
+    conf.projectPath = canonicalizePath(conf, getCurrentDir())
+    conf.projectIsStdin = true
+  elif conf.projectName != "":
+    try:
+      conf.projectFull = canonicalizePath(conf, conf.projectName)
+    except OSError:
+      conf.projectFull = conf.projectName
+    let p = splitFile(conf.projectFull)
+    let dir = if p.dir.len > 0: p.dir else: getCurrentDir()
+    conf.projectPath = canonicalizePath(conf, dir)
+    conf.projectName = p.name
+  else:
+    conf.projectPath = canonicalizePath(conf, getCurrentDir())
+
+proc loadConfigsAndRunMainCommand*(self: ProgBase, cache: IdentCache; conf: ConfigRef): bool =
+  loadConfigs(DefaultConfig, cache, conf) # load all config files
+  if self.suggestMode:
+    conf.command = "nimsuggest"
+
+  proc runNimScriptIfExists(scriptFile: string)=
+    if fileExists(scriptFile):
+      runNimScript(cache, scriptFile, freshDefines=false, conf)
+
+  # TODO:
```

Comment:
Don't write what's left TODO we have enough of this in the compiler, do it.

Reply:
done